[NVIDIA TF] Enable BF16 Conv Ops

kaixih opened the pull request:
This PR enables BF16 Conv ops via cuDNN or cuBLASLt.

This includes the fwd and bwd ops for the Conv2D/3D and the Depthwise Conv.

cc. @nluehr @pjannaty @kushanam

kaixih committed:
Support bf16 conv and add unit tests

Fix a potential issue of recursively calling

kaixih commented:
Rebased to clean up the commit history.

kushanam commented on tensorflow/core/util/gpu_device_functions.h:
```diff
@@ -725,6 +725,17 @@ __device__ inline Eigen::half GpuAtomicAdd(Eigen::half* ptr,
+        Eigen::bfloat16 ret =
+            Eigen::numext::bit_cast<Eigen::bfloat16>(a) + value;
+        return Eigen::numext::bit_cast<Eigen::half>(ret);
+      });
```
So I assume we can't do `GpuAtomicCasHelper` with bfloat16 which is the reason for the extra conversion?

kaixih replied:
Yes, GpuAtomicCasHelper only has the instance for fp16. Instantiating it for bf16 is not trivial so I am thinking about reusing the fp16 instance since both fp16 and bf16 have 16-bit float and the logic of atomic cas for them is same.

kaixih commented on tensorflow/core/kernels/depthwise_conv_op.h:
```diff
@@ -80,6 +84,29 @@ struct LaunchDepthwiseConvBackpropFilterOp {
+  bool operator()(OpKernelContext* ctx) const {
+    bool ret = false;
+    if (DataTypeToEnum<T>::value == DT_HALF) {
+      ret = true;
```
No, the original code path doesn't check the GPU arch. So, I follow it.

kaixih commented on tensorflow/core/kernels/conv_grad_input_ops.cc:
```diff
@@ -130,6 +130,16 @@ void LaunchConv2DBackpropInputOp<GPUDevice, T>::operator()(
+            "Conv2DBackpropInput for GPU with bfloat16 is only supported with "
+            "cuDNN on Ampere GPUs or later."));
+    return;
+  }
```
Why do I need another warning message, since we already have the error message?

reedwm commented:
Do you plan on supporting BF16 in all or most kernels so that it is usable outside XLA? I am concerned about the binary size increase that would be caused by instantiating many kernels with BF16, as well as the kernel maintenance cost.

kaixih commented:
Rebased the changes and addressed the suggestion. PTAL @kushanam

kaixih commented:
Just pushed a new commit to make the inclusion of bfloat16 optional with the build env var `TF_ENABLE_GPU_BFLOAT16` if binary size increase is really a concern here. FYI, the binary increase is about 3.8 MB after setting this env var for GPU architectures of major generations after sm 5.3+.

reedwm commented on tensorflow/python/kernel_tests/nn_ops/conv_ops_test.py:
```diff
@@ -1154,6 +1157,10 @@ def _RunAndVerifyBackpropFilter(self,
     for dtype in self._DtypesToTest(use_gpu=use_gpu):
+      # The backprop_filter may generate high err with a sequence of integers as
+      # the input (See _CreateNumpyTensor). So, we skip this dtype.
+      if dtype == dtypes.bfloat16:
```
We don't have any other backprop filter tests, which is problematic. Maybe it will pass if you round `expected` to the nearest bfloat16 value? If that still fails, maybe adding an `err_bf16` argument? (Note `assertAllCloseAccordingToType` takes a separate `bfloat16_rtol` arg.) Alternatively you can add a `skip_bf16` argument so you can disable only the failing tests.

kaixih replied:
Thanks for the feedback. I can try to check how far away the results are from the `expected` and play with the error/tolerance. Btw, can you point me to an example of `skip_bf16`? How such arg works?

reedwm replied:
I would start by rounding `expected` to the nearest bf16 value, by doing something like `expected = np.array(expected, dtype=tf.bfloat16.as_numpy_dtype)`. That might reduce the tolerance you need.

There is no existing example of `skip_bf16`. Basically, the idea is you add a `skip_bf16` argument to `_RunAndVerifyBackpropFilter` with a default value is False. Then, at this point in the code on line 1162, you do:

```
if dtype == dtypes.bfloat16 and skip_bf16:
  continue
```

That way, if the caller passes `skip_bf16=True`, the test is not run with bf16. Then, for tests that call `_RunAndVerifyBackpropFilter` but fail with bf16, you can pass `skip_bf16=True` to make the test pass.

kaixih replied:
Thanks for the suggestion. I just reran the tests and all of them are passed this time. So, I think the skip_bf16 is no longer needed. Removing this check.

reedwm commented on tensorflow/python/kernel_tests/nn_ops/depthwise_conv_op_base.py:
```diff
@@ -992,6 +993,10 @@ def testDepthwiseConv2DInputGradCompare(self):
         continue
```
The comment on 989 states ROCm doesn't support double, but now bf16 is also skipped. I don't know if bf16 is supported and whether this change allows TF to use it. If you don't know either, I'd add to the comment "It's unknown whether bf16 is supported".

And same for other places with this comment.

reedwm commented on tensorflow/python/kernel_tests/nn_ops/depthwise_conv_op_base.py:
```diff
@@ -954,25 +954,26 @@ def testDepthwiseConv2DFilterGradExplicit(self):
+    gpu_value = _GetVal(use_gpu=True, dtype=dtype)
+    cpu_value = _GetVal(use_gpu=False, dtype=dtype)
+    self.assertAllClose(cpu_value, gpu_value, rtol=rtol, atol=atol)
 
```
Can you also test bf16 in the functions that don't compare cpu and gpu? Namely:
* `testDepthwiseConv2DCudnn`
* `testDepthwiseConv2DExplicit`
* `testDepthwiseConv2DInputGradCudnn`
* `testDepthwiseConv2DInputGradExplicit`
* `testDepthwiseConv2DFilterGradCudnn`
* `testDepthwiseConv2DFilterGradExplicit`

This would be useful in case CPU and GPU both return the same incorrect results.

kaixih replied:
It seems all these tests are `run_v1_only` and are skipped when I tried to run them. Do you know any methods to run them in TF2 container?

reedwm replied:
I think setting the environment variable `TF2_BEHAVIOR=0` might allow them to run. Alternatively, passing `--config=v1` might work, but this might trigger a rebuild.

kaixih commented on tensorflow/core/kernels/conv_grad_filter_ops.cc:
```diff
@@ -725,6 +725,16 @@ void LaunchConv2DBackpropFilterOp<Eigen::GpuDevice, T>::operator()(
     return;
   }
 
+  if (DataTypeToEnum<T>::value == DT_BFLOAT16 &&
+      !stream->GetCudaComputeCapability().IsAtLeast(
+           se::CudaComputeCapability::AMPERE)) {
+    ctx->SetStatus(
+        errors::Unimplemented(
+            "Conv2DBackpropFilter for GPU with bfloat16 is only supported with "
+            "cuDNN on Ampere GPUs or later."));
+    return;
+  }
```
Let me sync with our cuDNN team on whether these checks can be relaxed.

reedwm replied:
Sounds good, let me know when you hear back from them.

kaixih commented on tensorflow/core/util/gpu_device_functions.h:
```diff
@@ -725,6 +725,17 @@ __device__ inline Eigen::half GpuAtomicAdd(Eigen::half* ptr,
+__device__ inline Eigen::bfloat16 GpuAtomicAdd(Eigen::bfloat16* ptr,
+                                               Eigen::bfloat16 value) {
+  Eigen::half ret = detail::GpuAtomicCasHelper(
+      reinterpret_cast<Eigen::half*>(ptr), [value](Eigen::half a) {
```
Yes, it seems we don't have to use the `std::launder` in this PR. But can you elaborate a little more about whey using `reinterpret_cast` here is undefined behavior? My understanding here is that no matter for fp16 or bf16, for this CAS operation, we only care about the bitwise comparison to swap or not. So, I feel this should be safe, right?

reedwm replied:
My impression is that using `reinterpret_cast` to interpret the bits of one type as another type is undefined behavior. E.g., the following is undefined:

```c++
float x = 123.
int y = *reinterpret_cast<int*>(&x);
```

In [this cppreference documentation](https://en.cppreference.com/w/cpp/language/reinterpret_cast), the "Type aliasing" section indicates this is not allowed. Deferencing a pointer generated with `reinterpret_cast` seems to be only allowed in very limited cases, such as when you `reinterpret_cast` to a `char*`.

To fix, we should instead use `memcpy` to copy interpret the bits of one type as another type. `Eigen::numext::bit_cast` internally uses a memcpy, so we should use that (on non-pointer types) instead of `reinterpret_cast`ing.

But since this is already an issue in the code and your PR's use of `reinterpret_cast` is consistent with the rest of the code, I wouldn't worry about it. Maybe if I have time I'll try fixing in a future change if it doesn't degrade performance.

kaixih commented:
> I think setting the environment variable TF2_BEHAVIOR=0 might allow them to run. Alternatively, passing --config=v1 might work, but this might trigger a rebuild.

I tried the following but with no luck and still saw the test was skipped.
```
TF2_BEHAVIOR=0 python depthwise_conv_op_test.py DepthwiseConv2DTest.testDepthwiseConv2DCudnn
```

For the alternative solution, do you mean I should try the bazel test instead of the python command like:
```
bazel test --config=v1 "//tensorflow/python/kernel_tests/nn_ops:depthwise_conv_op_test"
```

reedwm commented:
Presumably in your first command you're building the pip package, installing it, then running the test using `python` using the installed pip package. Try directly building and running with bazel instead:

```bash
bazel build //tensorflow/python/kernel_tests/nn_ops:depthwise_conv_op_test_gpu
TF2_BEHAVIOR=0 bazel-bin/tensorflow/python/kernel_tests/nn_ops/depthwise_conv_op_test_gpu DepthwiseConv2DTest.testDepthwiseConv2DCudnn
```

This is also a lot faster than rebuilding the pip package.

kaixih commented:
Thanks @reedwm . I have added more tests for the depthwise conv + bf16.

Regarding the bf16 support, our cudnn team still recommends the bf16 for the Ampere+, because we don't do an extensive tests over pre-Ampere and so there is no guarantee of function coverage. For cublas(lt), one can use bf16 for pre-Ampere but at the cost of runtime conversion overhead (it is essentially float computation). So, for all libraries, our recommendation is that using the bf16 on Ampere or later.

kaixih commented:
@reedwm Can you help check why the `feedback/copybara` failed?

kaixih commented:
Gentle ping. @reedwm

reedwm commented:
This failed to build internally. Am fixing. Will be merged soon.

reedwm commented:
An issue is that since the op requires Ampere or later, this breaks TF2 tests that previously ran BF16 ops and had pre-Ampere GPUs. Before, the BF16 ops would be placed on the CPU but now, they are placed on the GPU and fail with an error.

I think this is OK and that we should fix the tests, but I need to check internally with others first. I'll let you know what we decide.

For cublas in #58159, I think we should not require Ampere or above, since it works with cublas with a performance penalty. But since you said there is not very extensive testing of cuDNN on pre-Ampere hardware, I don't think we should allow pre-Ampere hardware for convolutions.

reedwm reviewed (changes requested):
> An issue is that since the op requires Ampere or later, this breaks TF2 tests that previously ran BF16 ops and had pre-Ampere GPUs. Before, the BF16 ops would be placed on the CPU but now, they are placed on the GPU and fail with an error.
>
> I think this is OK and that we should fix the tests, but I need to check internally with others first. I'll let you know what we decide.

I think the best way to resolve this is to support BF16 on pre-Ampere hardware, by casting the inputs to the convolution to FP32 and the outputs back to BF16. We already do this in the depthwise conv2d gradient [here](https://github.com/tensorflow/tensorflow/blob/ddea73d972eb5a66aca254faf1605656cee11456/tensorflow/core/kernels/depthwise_conv_grad_op.cc#L1192) for FP16. Altneratively, if you think cuDNN's support for BF16 on pre-Ampere hardware is good enough, you can use that.

For matmuls in #58159, just use cublas even on pre-Ampere hardware, since it's supported.

Both cases will be slow on pre-Ampere hardware. But this is better than crashing, since before bf16 previously worked by running on the CPU, which was likely even slower.

kaixih commented:
> this breaks TF2 tests that previously ran BF16 ops and had pre-Ampere GPUs. 

Can you elaborate a little more about the TF2 tests? Are they some of your internal tests?

Or are you talking about the unit tests in this PR?

For the conv ops, we didn't have the bf16 kernels before, so how did you test it on pre-ampere gpus? I think I am still a bit confused about why this PR would break the "TF2 tests".

@reedwm

kaixih commented:
It seems you'd like something of this logic in the kernel as below. Basically, we want bf16 support for all gpus. Am I getting it correct? @reedwm 

```
if dtype == bf16 and cc < 8.0
  input = cast(input, fp32)

conv(input, ...)

if dtype == bf16 and cc < 8.0
  output = cast(output, fp32)
```

reedwm commented:
The test that broke in [conv_ops_3d_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/nn_ops/conv_ops_3d_test.py). The test already uses bfloat16 [here](https://github.com/tensorflow/tensorflow/blob/860538f62384fc4d9427aeac6f20ff7a87bcd21b/tensorflow/python/kernel_tests/nn_ops/conv_ops_3d_test.py#L67), which is intended to run on the CPU using the bf16 CPU kernels. However, despite the fact bf16 is only used if `use_gpu` is False, the test will still use the GPU in TF2 if a GPU is available. TF2 is enabled in this test by setting the environment variable TF2_BEHAVIOR=1.

Fixing this particular test is easy, but I worry we may also break users. So I think the op should support bf16 on all GPUs.

> It seems you'd like something of this logic in the kernel as below. Basically, we want bf16 support for all gpus. Am I getting it correct?

Yes, that is correct. Note we do this in the depthwise conv2d gradient [here](https://github.com/tensorflow/tensorflow/blob/ddea73d972eb5a66aca254faf1605656cee11456/tensorflow/core/kernels/depthwise_conv_grad_op.cc#L1192) for FP16 (although without the compute capability check).

kaixih commented:
Added the support of bf16 on pre-ampere GPUs. Also, the bf16 tests are no longer skipped for pre-ampere GPUs.

Squashed the commits to better resolve the conflict.

PTAL. @reedwm

reedwm commented:
For some reason, this PR is mistakenly marked as closed, but it was actually merged in d787cdf78f3e9574a14b6d774ab2308a35228d9c.

teijeong commented:
I've attached some compile error messages from the internal builds. Hope that helps a bit on resolving the issue.

kaixih committed:
Update bazel BUILD

kaixih commented:
Reopening the PR.

From the log provided in https://github.com/tensorflow/tensorflow/issues/58480, it seems the linking issue is related to the newly added calls to `CastFunctor`, which are defined in `:cast_op`. So, adding it in the bazel BUILD file. (Not sure why my local build didn't hit this linking issue).

@teijeong PTAL. Thanks.

reedwm commented:
@kaixih, can you include the changes from the merge commit d787cdf78f3e9574a14b6d774ab2308a35228d9c? The merge commit made some extra changes to the PR to get some internal builds to work (although unfortunately, as we see with the rollback, not all internal builds).

I'm not sure what git commands to use to bring in the changes from the merge commit, but I can find out if you don't know.

kaixih committed:
Minor change

kaixih commented:
Yes, thanks for pointing it out. I cherry-picked this merging commit (https://github.com/tensorflow/tensorflow/commit/d787cdf78f3e9574a14b6d774ab2308a35228d9c). Also, @trevor-m pointed out a potential issue of recursively calling the specialized bfloat16 template from [here](https://github.com/tensorflow/tensorflow/blob/82570fd313cf51b3b70114b851108903b32c607a/tensorflow/core/kernels/conv_ops.cc#L1125) (my local build and the CI didn't hit any issue though). But I fixed this by separating out the logic to the `LaunchXXXImp` functions.

After that, I have to squash the commits to help resolve the conflicts (introduced a couple hours ago).

@reedwm PTAL.

kaixih commented:
Rebasing to resolve the conflict.
